spec/controllers/items_controller_spec.rb

```diff
@@ -0,0 +1,131 @@
+require "rails_helper"
+
+RSpec.describe ItemsController do
+  context "when user is logged in" do
+    before { sign_in_user }
+    let(:wishlist) { create(:wishlist, user: controller.current_user) }
+    let(:item) { create(:item, wishlist: wishlist) }
+
+    describe "GET #edit" do
+      context "when wishlist exists" do
+        context "when item exists" do
+          before { get_edit(item_id: item.id, wishlist_id: wishlist.id) }
+
+          it "returns 200 OK" do
+            expect(response).to have_http_status(:ok)
+          end
+
+          it "renders the 'edit' template" do
+            expect(response).to render_template(:edit)
+          end
+        end
+
+        context "when item doesn't exist" do
+          before { get_edit(item_id: -1, wishlist_id: wishlist.id) }
+
+          it "redirects to the root page" do
+            expect(response).to redirect_to(root_url)
+          end
+
+          it "shows an error" do
+            expect(flash[:alert]).not_to be_nil
+          end
+        end
+      end
+
+      context "when wishlist doesn't exist" do
+        before { get_edit(item_id: item.id, wishlist_id: -1) }
+
+        it "redirects to the root page" do
+          expect(response).to redirect_to(root_url)
+        end
+
+        it "shows an error" do
+          expect(flash[:alert]).not_to be_nil
+        end
+      end
+    end
+
+    describe "PUT #update" do
+      let(:new_params) { { name: "name" } }
+
+      context "when params valid" do
+        subject! { put_update(item.id, wishlist.id, new_params) }
+
+        it "changes item attributes" do
+          expect(Item.find(item.id).name).to eq(new_params[:name])
+        end
+
+        it "doesn't change number of items" do
+          expect { subject }.not_to change(Item, :count)
+        end
+
+        it "redirects to the wishlist page" do
+          expect(response).to redirect_to(wishlist)
+        end
+
+        it "shows a notice" do
+          expect(flash[:notice]).not_to be_nil
+        end
+      end
+
+      context "when params invalid" do
+        before { put_update(item.id, wishlist.id, { name: nil }) }
+
+        it "re-renders 'edit' form" do
+          expect(response).to render_template(:edit)
+        end
+      end
+    end
+  end
+
+  context "when not logged in" do
+    let(:wishlist) { create(:wishlist) }
+    let(:item) { create(:item) }
+
+    shared_examples "redirects to the 'sign_in' page" do
+      it "redirects to the 'sign_in' page" do
+        expect(response).to redirect_to(new_user_session_path)
+      end
+    end
+
+    describe "GET #edit" do
```

Comment:
for me it should be

```
describe #GET
  context when user is logged in
    assertions
  context when user is logged out
    assertions
```

instead of

```
context when user is logged in
  describe #GET
    assertions
context when user is logged out
  describe #GET
    assertions
```

specs for a single method should be always under the same context (describe). If you will need to change the code in the future, you there will be a single place to update relevant specs.